lib/Controller/UsersController.php

```diff
@@ -469,6 +501,254 @@ public function create($username, $password, array $groups= [], $email='') {
+	 * @NoSubadminRequired
+	 *
+	 * @param string $token
+	 * @param string $userId
+	 * @return TemplateResponse
+	 */
+	public function setPasswordForm($token, $userId) {
+		try {
+			$this->checkPasswordSetToken($token, $userId);
+		} catch (UserTokenException $e) {
+			if ($e->getCode() === UserTokenException::TOKEN_TIME_EXPIRED) {
+				return new TemplateResponse(
+					'user_management', 'new_user/resendtokenbymail',
+					[
+						'link' => $this->urlGenerator->linkToRouteAbsolute('user_management.Users.resendToken', ['userId' => $userId])
+					], 'guest'
+				);
+			}
+			$this->log->logException($e, ['app' => 'user_management']);
+			$tokenMessage = null;
+			if ($e->getCode() === UserTokenException::TOKEN_AND_TIME_INVALID) {
+				$tokenMessage = $this->l10n->t('The token provided is invalid.');
+			} else {
+				$tokenMessage = $this->l10n->t('The token mismatch happened.');
+			}
+			return new TemplateResponse(
+				'core', 'error',
+				[
+					"errors" => [["error" => $tokenMessage]]
+				], 'guest'
+			);
+		}
+
+		return new TemplateResponse(
+			'user_management', 'new_user/setpassword',
+			[
+				'link' => $this->urlGenerator->linkToRouteAbsolute('user_management.Users.setPassword', ['userId' => $userId, 'token' => $token])
+			], 'guest'
+		);
+	}
+
+	/**
+	 * @param string $token
+	 * @param string $userId
+	 * @return null
+	 * @throws UserTokenException
+	 */
+	private function checkPasswordSetToken($token, $userId) {
+		$user = $this->userManager->get($userId);
+
+		$splittedToken = \explode(':', $this->config->getUserValue($userId, 'owncloud', 'lostpassword', null));
+		if (\count($splittedToken) !== 2) {
+			$this->config->deleteUserValue($userId, 'owncloud', 'lostpassword');
+			throw new UserTokenException(UserTokenException::TOKEN_AND_TIME_INVALID);
+		}
+
+		if ($splittedToken[0] < ($this->timeFactory->getTime() - 60*60*12) ||
+			$user->getLastLogin() > $splittedToken[0]) {
+			$this->config->deleteUserValue($userId, 'owncloud', 'lostpassword');
+			throw new UserTokenException(UserTokenException::TOKEN_TIME_EXPIRED);
+		}
+
+		if (!\hash_equals($splittedToken[1], $token)) {
+			$this->config->deleteUserValue($userId, 'owncloud', 'lostpassword');
+			throw new UserTokenException(UserTokenException::TOKEN_MISMACTH);
+		}
+	}
+
+	/**
+	 * @PublicPage
+	 * @NoCSRFRequired
+	 * @NoAdminRequired
+	 * @NoSubadminRequired
+	 * @param $userId
+	 */
+	public function resendToken($userId) {
+		$user = $this->userManager->get($userId);
+
+		if ($user === null) {
+			return new TemplateResponse(
+				'core', 'error',
+				[
+					"errors" => [["error" => $this->l10n->t('User: %s does not exist', [$userId])]]
+				],
+				'guest'
+			);
+		}
+
+		if ($user->getEMailAddress() === null) {
+			return new TemplateResponse(
+				'core', 'error',
+				[
+					"errors" => [["error" => $this->l10n->t('Email address not set for: %s', [$userId])]]
+				],
+				'guest'
+			);
+		}
+
+		$token = $this->secureRandom->generate(21,
```

Comment:
don't we have such logic elsewhere already for the original token creation and sending ? if yes, please refactor into a common method

Reply:
Done. Moved the common logic to a separate method.